### PR TITLE
[BOLT] Use symbol table info in registerFragment

### PR DESCRIPTION
**Pull request by aaupov**
Fragment matching relies on symbol names to identify and register split
function fragments. However, as split fragments are often local symbols,
name aliasing is possible. For such cases, use symbol table to resolve
ambiguities.

This requires the presence of FILE symbols in the input binary. As BOLT
requires non-stripped binary, this is a reasonable assumption. Note that
`strip -g` removes FILE symbols by default, but `--keep-file-symbols`
can be used to preserve them.

Depends on: https://github.com/llvm/llvm-project/pull/89861

Test Plan:
Updated X86/fragment-lite.s

### CODE REVIEW

File: bolt/include/bolt/Rewrite/RewriteInstance.h [hunk 1]
```diff
@@ -490,6 +490,9 @@ class RewriteInstance {
   /// Store all non-zero symbols in this map for a quick address lookup.
   std::map<uint64_t, llvm::object::SymbolRef> FileSymRefs;
 
+  /// FILE symbols.
+  std::vector<object::DataRefImpl> FileSymbols;
```
Comment (dcci) [inline bolt/include/bolt/Rewrite/RewriteInstance.h:494]: This comment is perfunctory. You might want to consider making it more explicit or just remove it altogether.

File: bolt/lib/Rewrite/RewriteInstance.cpp [hunk 2]
```diff
@@ -1417,50 +1418,106 @@ void RewriteInstance::registerFragments() {
...
+        BD = BC->getBinaryDataByName((Twine(ParentName) + "/1").str());
+      } else if (BD && (!NumPossibleLocalParents || Suffix.empty())) {
+        // Global parent and either no local candidates (second most common), or
+        // the fragment is global as well (uncommon).
+      } else {
+        // Any other case: need to disambiguate using FILE symbols.
+        AmbiguousFragments.emplace_back(ParentName, &Function);
         continue;
       }
+      assert(BD);
```
Comment (dcci) [inline bolt/lib/Rewrite/RewriteInstance.cpp:1447]: add a message?

Comment (dcci) [reply bolt/lib/Rewrite/RewriteInstance.cpp:1447]: Also, why is this an assertion and not an `exit(1)`-style error?

File: bolt/lib/Rewrite/RewriteInstance.cpp [hunk 3]
```diff
@@ -1417,50 +1418,106 @@ void RewriteInstance::registerFragments() {
...
+      if (cantFail(Symbol.getName()) == ParentName) {
+        ParentAddress = cantFail(Symbol.getAddress());
+        break;
+      }
+    }
+
+    // No local parent is found, use global parent function.
+    if (!ParentAddress) {
+      BinaryData *ParentBD = BC->getBinaryDataByName(ParentName);
+      assert(ParentBD);
```
Comment (dcci) [inline bolt/lib/Rewrite/RewriteInstance.cpp:1513]: ditto

File: bolt/lib/Rewrite/RewriteInstance.cpp [hunk 4]
```diff
@@ -1417,50 +1418,106 @@ void RewriteInstance::registerFragments() {
...
+      ParentAddress = ParentBD->getAddress();
+    }
+
+    BinaryFunction *ParentBF = BC->getBinaryFunctionAtAddress(ParentAddress);
+    assert(ParentBF);
```
Comment (dcci) [inline bolt/lib/Rewrite/RewriteInstance.cpp:1518]: ditto

**Review by dcci — CHANGES_REQUESTED** [review]

File: bolt/lib/Rewrite/RewriteInstance.cpp (hunk 2 above)
Comment (aaupov) [reply bolt/lib/Rewrite/RewriteInstance.cpp:1446]: I'm on the fence regarding making it an error over an assertion.
Assertions should be used for internal invariants, whereas exit(1)/errors should be used for validating input. The absence of a parent fragment is the latter and that's why I used exit(1) in BinaryFunction lookup. I guess I should treat BD similarly because BD->getAddress would crash in Release build mode.

**Review by dcci — APPROVED** [review]

File: bolt/lib/Rewrite/RewriteInstance.cpp [hunk 5]
```diff
@@ -1417,50 +1418,108 @@ void RewriteInstance::registerFragments() {
...
-          BD = BDLocal;
-      }
-      if (!BD) {
-        if (opts::Verbosity >= 1)
-          BC->outs() << "BOLT-INFO: parent function not found for " << Name
-                     << "\n";
+      uint64_t NumPossibleLocalParents = NR.getNumDuplicates(ParentName);
+      // The most common case: single local parent fragment.
+      if (!BD && NumPossibleLocalParents == 1) {
+        BD = BC->getBinaryDataByName((Twine(ParentName) + "/1").str());
```
Comment (maksfb) [inline bolt/lib/Rewrite/RewriteInstance.cpp:1439]: Please use `NameResolver` interface.

File: bolt/include/bolt/Rewrite/RewriteInstance.h [hunk 6]
```diff
@@ -490,6 +490,9 @@ class RewriteInstance {
   /// Store all non-zero symbols in this map for a quick address lookup.
   std::map<uint64_t, llvm::object::SymbolRef> FileSymRefs;
 
+  /// FILE symbols used for disambiguating split function parents.
+  std::vector<object::DataRefImpl> FileSymbols;
```
[the remark this answers is not in the document]
Comment (aaupov) [reply bolt/include/bolt/Rewrite/RewriteInstance.h:494]: Yes. SymbolRef contains DataRefImpl and an object pointer which in our case is always `InputFile`.

File: bolt/include/bolt/Utils/NameResolver.h [hunk 7]
```diff
@@ -28,6 +28,13 @@ class NameResolver {
   static constexpr char Sep = '/';
 
 public:
+  /// Return the number of duplicates registered for a given \p Name.
```
[the remark this answers is not in the document]
Comment (aaupov) [reply bolt/include/bolt/Utils/NameResolver.h:31]: Let me rename/rephrase it to "counter" instead to make it more clear.
I took "duplicates" from the comment above.

File: bolt/include/bolt/Rewrite/RewriteInstance.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (maksfb) [reply bolt/include/bolt/Rewrite/RewriteInstance.h:494]: If memory is a concern, you should free it once done processing. I find the savings questionable to be worth it over readability, especially when you compare symbols.

File: bolt/lib/Rewrite/RewriteInstance.cpp [hunk 8]
```diff
@@ -1417,50 +1418,125 @@ void RewriteInstance::registerFragments() {
...
-      const uint64_t Address = BD->getAddress();
-      BinaryFunction *BF = BC->getBinaryFunctionAtAddress(Address);
-      if (!BF) {
-        if (opts::Verbosity >= 1)
-          BC->outs() << formatv(
-              "BOLT-INFO: parent function not found at {0:x}\n", Address);
-        continue;
+      if (BD) {
+        const uint64_t Address = BD->getAddress();
+        BinaryFunction *BF = BC->getBinaryFunctionAtAddress(Address);
```
Comment (maksfb) [inline bolt/lib/Rewrite/RewriteInstance.cpp:1450]:
```suggestion
        BinaryFunction *BF = BC->getFunctionForSymbol(BD->getSymbol());
```

File: bolt/include/bolt/Utils/NameResolver.h [hunk 9]
```diff
@@ -28,10 +28,23 @@ class NameResolver {
   static constexpr char Sep = '/';
 
 public:
+  /// Return the counter for a given \p Name.
+  uint64_t getCounter(StringRef Name) const {
```
Comment (maksfb) [inline bolt/include/bolt/Utils/NameResolver.h:32]: `getUniquifiedNameCount()`?

File: bolt/include/bolt/Utils/NameResolver.h [hunk 10]
```diff
@@ -28,10 +28,23 @@ class NameResolver {
...
+    if (Counters.contains(Name))
+      return Counters.at(Name);
+    return 0;
+  }
+
   /// Return unique version of the \p Name in the form "Name<Sep><Number>".
+  std::string getUniqueName(StringRef Name, const uint64_t ID) const {
```
Comment (maksfb) [inline bolt/include/bolt/Utils/NameResolver.h:39]: Update the commend above with `ID`.

File: bolt/lib/Rewrite/RewriteInstance.cpp [hunk 11]
```diff
@@ -1417,50 +1418,125 @@ void RewriteInstance::registerFragments() {
...
+  std::vector<std::pair<StringRef, BinaryFunction *>> AmbiguousFragments;
   for (auto &BFI : BC->getBinaryFunctions()) {
     BinaryFunction &Function = BFI.second;
     if (!Function.isFragment())
       continue;
-    unsigned ParentsFound = 0;
     for (StringRef Name : Function.getNames()) {
       StringRef BaseName, Suffix;
       std::tie(BaseName, Suffix) = Name.split('/');
       const size_t ColdSuffixPos = BaseName.find(".cold");
```
[the remark this answers is not in the document]
Comment (aaupov) [reply bolt/lib/Rewrite/RewriteInstance.cpp:1431]: Suffix is used to determine if a given fragment name is a local symbol. How can I check it otherwise?

[the remark this answers is not in the document]
Comment (maksfb) [reply bolt/lib/Rewrite/RewriteInstance.cpp:1431]: `BaseName == Name`?

[the remark this answers is not in the document]
Comment (aaupov) [reply bolt/lib/Rewrite/RewriteInstance.cpp:1431]: Oh right, I used a similar check just recently in an [alternative implementation](https://github.com/llvm/llvm-project/pull/89320/files#diff-4be6195ccdb53447f870b9b26a36d0b916ace3cbd6efe595bb3cbffd53cf0dacR1441).

**Comment by aaupov** [comment]
@maksfb: I've addressed the existing concerns – does that look good for you? Note we now depend on #89861 for ELFSymbolRef comparisons. We can add heuristics based on bolt_pseudo.o to make fragment matching even faster in BOLTed binary case as follow-up diffs.

**Review by maksfb — COMMENTED** [review]
Otherwise LGTM.

File: bolt/lib/Rewrite/RewriteInstance.cpp [hunk 12]
```diff
@@ -1417,50 +1418,125 @@ void RewriteInstance::registerFragments() {
...
+  uint32_t FirstGlobal{0};
+  auto *ELF64LEFile = cast<ELF64LEObjectFile>(InputFile);
+  const ELFFile<ELF64LE> &Obj = ELF64LEFile->getELFFile();
+  for (const auto &Sec : cantFail(Obj.sections())) {
+    if (Sec.sh_type == ELF::SHT_SYMTAB) {
+      FirstGlobal = Sec.sh_info;
+      break;
+    }
+  }
+  if (FirstGlobal == 0) {
```
[the remark this answers is not in the document]
Comment (aaupov) [reply bolt/lib/Rewrite/RewriteInstance.cpp]: I've re-read the documentation carefully and it says that sh_info has the value one larger than the last local symbol index: https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-94076.html#scrolltoc (Table 12-9 ELF sh_link and sh_info Interpretation).
So I'm dropping the check and exit.

**Review by maksfb — APPROVED** [review]
